File: packages/dart/lib/src/crypto_provider.dart
import 'dart:typed_data';

abstract class CkvfCrypto {
  Uint8List randomBytes(int n);

  Future<Uint8List> sha256(List<int> data);

  Future<({Uint8List ciphertext, Uint8List tag})> aes256gcmEncrypt(
    List<int> key,
    List<int> iv,
    List<int> plaintext,
    List<int> aad,
  );

  Future<Uint8List> aes256gcmDecrypt(
    List<int> key,
    List<int> iv,
    List<int> ciphertext,
    List<int> tag,
    List<int> aad,
  );

  Future<Uint8List> argon2id({
    required List<int> password,
    required List<int> salt,
    required int m,
    required int t,
    required int p,
    required int keyLength,
  });

  Future<({Uint8List publicKey, Uint8List privateKey})> ed25519Generate();

  Future<Uint8List> ed25519PublicFromSeed(List<int> seed);

  Future<Uint8List> ed25519Sign(List<int> seed, List<int> message);

  Future<bool> ed25519Verify(
    List<int> publicKey,
    List<int> message,
    List<int> signature,
  );
}

bool constantTimeEqual(List<int> a, List<int> b) {
  if (a.length != b.length) return false;
  var diff = 0;
  for (var i = 0; i < a.length; i++) {
    diff |= a[i] ^ b[i];
  }
  return diff == 0;
}

File: packages/js/packages/core/src/crypto-provider.ts
export interface CkvfCrypto {
  randomBytes(n: number): Uint8Array;
  sha256(data: Uint8Array): Promise<Uint8Array>;
  aes256gcmEncrypt(
    key: Uint8Array,
    iv: Uint8Array,
    plaintext: Uint8Array,
    aad: Uint8Array,
  ): Promise<{ ciphertext: Uint8Array; tag: Uint8Array }>;
  aes256gcmDecrypt(
    key: Uint8Array,
    iv: Uint8Array,
    ciphertext: Uint8Array,
    tag: Uint8Array,
    aad: Uint8Array,
  ): Promise<Uint8Array>;
  argon2id(params: {
    password: Uint8Array;
    salt: Uint8Array;
    m: number;
    t: number;
    p: number;
    keyLength: number;
  }): Promise<Uint8Array>;
  ed25519Generate(): Promise<{ publicKey: Uint8Array; privateKey: Uint8Array }>;
  ed25519PublicFromSeed(seed: Uint8Array): Promise<Uint8Array>;
  ed25519Sign(seed: Uint8Array, message: Uint8Array): Promise<Uint8Array>;
  ed25519Verify(publicKey: Uint8Array, message: Uint8Array, signature: Uint8Array): Promise<boolean>;
}

export function constantTimeEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) diff |= a[i]! ^ b[i]!;
  return diff === 0;
}
